### ISSUE TITLE
v 2.0 1256 doesn't start on mac

### ISSUE THREAD

**Issue by mistermax80** [akasolace/HO#166]
![Schermata 2019-07-06 alle 08 40 13](https://user-images.githubusercontent.com/15519907/60752775-f2aa4a00-9fc9-11e9-8a30-6b1cd0ffd003.png)
[HO-2019-07-06.log](https://github.com/akasolace/HO/files/3364480/HO-2019-07-06.log)

**Comment by akasolace** [comment]
08:38:56 [Error]   core.model.FormulaFactors: Error while loading prediction/defaults.xml
Under mac prediction/defaults.xml is not found. I had to change that part recently, I tested it with zip, and windows exe but not with mac.
@mistermax80 can you do a step by step debug and tell me more about it ?

**Comment by akasolace** [comment]
@mistermax80  I do not have access to a mac anymore so I really rely on you for this one.
I can't do it myself but I am happy to give you advice if you are struggling.

**Comment by akasolace** [comment]
@pierpaolocira  maybe you can help @mistermax80  on this one.
One way to start is for example to look at how the prediction files are loaded and do the same within a ifmac clause.

**Comment by pierpaolocira** [comment]
@akasolace OK, I should be able to take a look in some days... but I have some trubles in reproducing.
(Maybe stupid questions, but I have no experience both with this project and gradle - I'm a mvn fan :) )

From the screenshot it seems that it happens at startup from a compiled version, but the `:createDmg` task fails to me if I launch `gradle build`.
Indeed, if I launch it from the IDE I have no troubles.

Should I do something specific in order to build? I have a:
```
Execution failed for task ':createDmg'.
> Unable to determine image size with sip
```
I think there is something wrong with the path, because it searches the image file in
`/path/to/project/HO-pierpaolocira/build/macApp/.background/HO-pierpaolocira\buildResources\Mac\Back.png`
instead of
`/path/to/project/HO-pierpaolocira/build/macApp/.background/Back.png`

I launch the command from: `/path/to/project/HO-pierpaolocira/`.


Or, maybe, there is a quick way to reproduce by launching from the IDE?

**Comment by akasolace** [comment]
@pierpaolocira Indeed from the IDE it will work but fails from the app because it will look the ressource in the home folder instead of looking at it within the prediction folder within the jar.

I don't think building the app will help in closing this issue, however if you want to build it, I suggest you create a minimal task with only the necessary steps. maye something like (I cant' test) :

task makeMacAPP(type: GradleBuild) {
    group 'ho'
    description 'Create MacOS binary'
    tasks = ['cleanBuild',  'installDist', 'AddResourcesInDistrib', 'ZipDistribution',  'osXjar', 'createApp', 'AddResourcesAndZipMacApp',] }

**Comment by akasolace** [comment]
> I think there is something wrong with the path, because it searches the image file in
/path/to/project/HO-pierpaolocira/build/macApp/.background/HO-pierpaolocira\buildResources\Mac\Back.png

That path looks ok to me. Does it not exist on your system ?

**Comment by akasolace** [comment]
@pierpaolocira you can join the dev team on slack  
https://join.slack.com/t/ho-tle6272/shared_invite/enQtNzAxMzU2ODEzODg5LWY5OTVmYTY1OGI3MjZmZjUwYTRlYmJlZjMzN2Q2MTE1MWRhNTVlNjBhYjAwNmE1NjQxYjc2NGMwNDhkNmU4MjQ

**Comment by pierpaolocira** [comment]
@akasolace 
> Indeed from the IDE it will work but fails from the app
> I don't think building the app will help in closing this issue
I think this is the only way I have to test either the fix works or not... or do you see other possibilities?


>  if you want to build it, I suggest you create a minimal task with only the necessary steps.
Thanks, I'll try this way... we'll investigate in a second moment why it doesn't perform the full build on Mac (at least on mine :) )

**Comment by akasolace** [comment]
@pierpaolocira  the full build would require nsis to create the windows .exe, maybe you do not have it ...

**Comment by pierpaolocira** [comment]
Oh, I misunderstood your initial reply... sorry

**Comment by pierpaolocira** [comment]
@akasolace I'm finally able to build the OSX executable and reproduce, but I had to change the build.gradle by commenting the line `bundleJRE = true` into `macAppBundle {`

Indeed, without it, Java11 is embedded into the application, and the source code is not yet compatible, so the application just shows a generic "JRE Error" in the splash screen.

Perhaps this relate to my workstation having Java up to 11... 
I like the idea to embed a JRE (I think it follows the Oracle guidelines related to distributing desktop applications, in order to avoid compatibility issues), but I think this should be changed until Java 11 is not supported... should I open an issue for this?

**Comment by pierpaolocira** [comment]
@akasolace @mistermax80 
I'm quite confident that there is a configuration issue in building (but I repeat, I'm quite new with gradle), because I cannot find any "prediction" folder either in the HO.app file or in the related HO.jar

A few questions:
- is there any reason not to invoke `osXjar` into the `makeosX` task? I think it should be added
- is there any reason to `exclude "prediction"` into the `osXjar` task?
Indeed if I invoke also `osXjar` in the chain of tasks for building the app AND I comment that exclude, this issue doesn't happen.
(Please note that it will then fail in `Player.java`, most probably because `FormulaFactors.importDefaults()` try to load `prediction/defaults.xml` in a different way than `FormulaFactors.readFromXML()`).

Anyway, I'm not sure that commenting that "exclude" is the most appropriate way to fix. There should be a reason if it's there, so maybe "predictions" should be later included somewhere else?

(If we consider what I wrote, my previous comment, and also https://github.com/akasolace/HO/issues/182 I think that the full configuration file might need to be reviewed)

**Comment by akasolace** [comment]
@pierpaolocira 

> is there any reason not to invoke osXjar into the makeosX task? I think it should be added
yes, it should be added ( I just did)

> is there any reason to exclude "prediction" into the osXjar task?
Indeed if I invoke also osXjar in the chain of tasks for building the app AND I comment that exclude, this issue doesn't happen.

The idea was to separate was goes into the jar and was is easily accessible to user (for example a user can manually edit prediction formula file and for this it should not have to enter into the jar).

> There should be a reason if it's there, so maybe "predictions" should be later included somewhere else?

Instead they go into lib\ho\,  you can see that into ospackage task


I am familliar with windows and linux but not with osX as I have lost my VM and Iwould rather avoid to go again through that.

Can you please tell me in osX what goes into ~ and also the jar structure. It is working for all other files so I think we are missing something obvious here.

**Comment by pierpaolocira** [comment]
> Instead they go into lib\ho, you can see that into ospackage task

They **should** go... As written,  I cannot find any "prediction" folder either in the HO.app file or in the related HO.jar.

I generated them via the following tasks
`['init', 'installDist', 'AddResourcesInDistrib', 'ZipDistribution', 'osXjar', 'createApp', 'AddResourcesAndZipMacApp']`


> Can you please tell me in osX what goes into ~ and also the jar structure. It is working for all other files so I think we are missing something obvious here.

Sure.

Content of the generated HO-*.app
```
.
├── Contents
│   ├── Info.plist
│   ├── Java
│   │   ├── HO.jar
│   │   ├── animal-sniffer-annotations-1.17.jar
│   │   ├── checker-qual-2.5.2.jar
│   │   ├── commons-codec-1.4.jar
│   │   ├── error_prone_annotations-2.2.0.jar
│   │   ├── failureaccess-1.0.1.jar
│   │   ├── gson-2.8.5.jar
│   │   ├── guava-27.1-jre.jar
│   │   ├── hsqldb-2.4.1.jar
│   │   ├── j2objc-annotations-1.1.jar
│   │   ├── jcalendar-1.4.jar
│   │   ├── jgoodies-common-1.8.1.jar
│   │   ├── jgoodies-looks-2.7.0.jar
│   │   ├── jsr305-3.0.2.jar
│   │   ├── listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
│   │   └── scribe-1.3.0.jar
│   ├── MacOS
│   │   └── JavaAppLauncher
│   ├── PkgInfo
│   └── Resources
│       └── HO.icns
└── changelog.html
```

Content of HO.jar (only directories)
```
├── META-INF
├── core
│   ├── constants
│   │   └── player
│   ├── datatype
│   ├── db
│   │   ├── backup
│   │   └── frontend
│   ├── epv
│   ├── file
│   │   ├── hrf
│   │   └── xml
│   ├── gui
│   │   ├── comp
│   │   │   ├── entry
│   │   │   ├── panel
│   │   │   ├── renderer
│   │   │   ├── tabbedPane
│   │   │   └── table
│   │   ├── language
│   │   ├── model
│   │   ├── print
│   │   └── theme
│   │       ├── ho
│   │       ├── jgoodies
│   │       └── nimbus
│   ├── model
│   │   ├── cup
│   │   ├── match
│   │   ├── misc
│   │   ├── player
│   │   └── series
│   ├── module
│   │   └── config
│   ├── net
│   │   ├── login
│   │   └── test
│   ├── option
│   │   └── db
│   ├── prediction
│   │   └── engine
│   ├── rating
│   ├── training
│   │   └── type
│   └── util
├── flags
├── gui
│   └── bilder
│       ├── Fugue
│       ├── arrows
│       ├── highlights
│       ├── smilies
│       └── trainpreview
├── module
│   ├── evilcard
│   │   └── gui
│   ├── ifa
│   │   ├── config
│   │   ├── gif
│   │   ├── menu
│   │   ├── model
│   │   └── table
│   ├── lineup
│   │   ├── exchange
│   │   ├── penalties
│   │   └── substitution
│   │       ├── model
│   │       ├── plausibility
│   │       └── positionchooser
│   ├── matches
│   │   └── statistics
│   ├── matchesanalyzer
│   │   ├── data
│   │   └── ui
│   │       ├── cbox
│   │       └── table
│   │           └── cell
│   │               └── content
│   ├── misc
│   ├── nthrf
│   ├── opponentspy
│   │   └── OpponentTeam
│   ├── playerOverview
│   ├── playeranalysis
│   │   ├── experience
│   │   └── skillCompare
│   ├── pluginFeedback
│   ├── series
│   ├── specialEvents
│   │   ├── filter
│   │   └── table
│   ├── statistics
│   ├── teamAnalyzer
│   │   ├── ht
│   │   ├── manager
│   │   ├── report
│   │   ├── ui
│   │   │   ├── component
│   │   │   ├── controller
│   │   │   ├── lineup
│   │   │   ├── model
│   │   │   └── renderer
│   │   └── vo
│   ├── teamOfTheWeek
│   │   └── gui
│   ├── training
│   │   └── ui
│   │       ├── comp
│   │       ├── model
│   │       └── renderer
│   ├── transfer
│   │   ├── history
│   │   ├── scout
│   │   ├── test
│   │   ├── transfertype
│   │   └── ui
│   │       ├── layout
│   │       └── sorter
│   └── tsforecast
├── sprache
└── tool
    ├── arenasizer
    ├── dbcleanup
    ├── export
    ├── hrfExplorer
    ├── injury
    ├── keepertool
    ├── notepad
    └── updater
```

**Comment by akasolace** [comment]
@pierpaolocira  ok indeed those folders were incorrectly excluded.
I put them back, and I recreated artefacts, please let me know if it fixes the issue.
Thanks

**Comment by pierpaolocira** [comment]
> @pierpaolocira ok indeed those folders were incorrectly excluded.
> I put them back, and I recreated artefacts, please let me know if it fixes the issue.
> Thanks

@akasolace this was exactly the test I did some days ago. This partially solves the situation, because the loading moves forward until throwing another exception that seems to be related to the current issue.

Here the new logs are:
```
java.lang.NullPointerException
	at core.model.player.Player.getIdealPosition(Player.java:837)
	at core.gui.model.UserColumnFactory$27.getTableEntry(UserColumnFactory.java:750)
	at core.gui.model.PlayerOverviewModel.initData(PlayerOverviewModel.java:163)
	at core.gui.model.PlayerOverviewModel.setValues(PlayerOverviewModel.java:89)
	at module.playerOverview.PlayerOverviewTable.initModel(PlayerOverviewTable.java:144)
	at module.playerOverview.PlayerOverviewTable.<init>(PlayerOverviewTable.java:33)
	at module.playerOverview.SpielerUebersichtsPanel.initSpielerTabelle(SpielerUebersichtsPanel.java:182)
	at module.playerOverview.SpielerUebersichtsPanel.initComponents(SpielerUebersichtsPanel.java:121)
	at module.playerOverview.SpielerUebersichtsPanel.<init>(SpielerUebersichtsPanel.java:45)
	at module.playerOverview.PlayerOverviewModule.createTabPanel(PlayerOverviewModule.java:30)
	at core.gui.comp.tabbedPane.HOTabbedPane.showTab(HOTabbedPane.java:38)
	at core.gui.HOMainFrame.initComponents(HOMainFrame.java:494)
	at core.gui.HOMainFrame.<init>(HOMainFrame.java:188)
	at core.gui.HOMainFrame.instance(HOMainFrame.java:282)
	at core.HO$1.run(HO.java:225)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

Furthermore this puts back the files into the HO.jar, so this can't match with the idea to keep user editable files out of the JAR in order to allow an easy editing... :(

If you want I can try to understand how to solve it and to keep the MacOSX packaging coherent with the others... or maybe we can fix it now  and then have a separate task for packaging it properly...
The latter option would allow not to have this blocker in MacOSX for v2, but at the end we might have a bigger effort to achieve the desired result... I imagine this depends by the schedule we have for v2...

(For future reference, I imagine that this issue has been introduced around the commit of 10 weeks ago: f46137b793ef3944ac78559d06ce50f1de077cf0 )

**Comment by akasolace** [comment]
> Furthermore this puts back the files into the HO.jar, so this can't match with the idea to keep user editable files out of the JAR in order to allow an easy editing... :(

it appears It was actually never (easily) possible under mac,  does not look like a big deal to me, what do you think. Also I do not know how to deal with it differently

**Comment by akasolace** [comment]
> forward until throwing another exception that seems to be related to the current issue

I guess it fails within function readFromXML within FormulaFactors.java . Can you confirm and see if you can find where it crashes. I guess we need a specific call if platform is osX, see how dat files are read for examples.
readFromFile calls () in RatingPredictionConfig.java might help

**Comment by akasolace** [comment]
> If you want I can try to understand how to solve it and to keep the MacOSX packaging coherent with the others... or maybe we can fix it now and then have a separate task for packaging it properly...
The latter option would allow not to have this blocker in MacOSX for v2, but at the end we might have a bigger effort to achieve the desired result... I imagine this depends by the schedule we have for v2...


I am not sure of what you mean but for clarity we will of course not release the v2 until it works under mac. 
Now going forward if you find a way to store prediction outside of the JAR will be nice but in my opinion it is really not a priority. If you consider the % of peaople under mac and the % of those who will manually edit the prediction files I would personally not really bother ....

**Comment by pierpaolocira** [comment]
@akasolace  I just put back the "exclude", built the zip via `gradle clean installDist AddResourcesInDistrib ZipDistribution` (I don't know if I run more tasks than necessary), and directly run via `java -jar HO.jar`.

And it works.

Please take a look to the structure of the folder (`prediction` is at the same level of the jars).
```
├── HO.bat
├── HO.jar
├── Logo.ico
├── animal-sniffer-annotations-1.17.jar
├── changelog.html
├── checker-qual-2.5.2.jar
├── commons-codec-1.4.jar
├── db
│   ├── database.data
│   ├── database.properties
│   └── database.script
├── error_prone_annotations-2.2.0.jar
├── failureaccess-1.0.1.jar
├── gson-2.8.5.jar
├── guava-27.1-jre.jar
├── ho.sh
├── hsqldb-2.4.1.jar
├── j2objc-annotations-1.1.jar
├── jcalendar-1.4.jar
├── jgoodies-common-1.8.1.jar
├── jgoodies-looks-2.7.0.jar
├── jsr305-3.0.2.jar
├── listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
├── logs
│   └── HO-2019-08-15.log
├── prediction
│   ├── default
│   │   ├── centralattack.dat
│   │   ├── centraldefense.dat
│   │   ├── midfield.dat
│   │   ├── playerstrength.dat
│   │   ├── sideattack.dat
│   │   ├── sidedefense.dat
│   │   └── tactics.dat
│   ├── defaults.xml
│   ├── old\ Engine
│   │   ├── centralattack.dat
│   │   ├── centraldefense.dat
│   │   ├── midfield.dat
│   │   ├── playerstrength.dat
│   │   ├── sideattack.dat
│   │   ├── sidedefense.dat
│   │   └── tactics.dat
│   ├── predictionTypes.conf
│   └── skilldrops
│       ├── README.txt
│       ├── defending
│       ├── keeper
│       ├── passing
│       ├── playmaking
│       ├── scoring
│       ├── setpieces
│       └── winger
├── scribe-1.3.0.jar
├── themes
└── user.xml
```

So, it should not be impossible.

But I'm mainly JEE developer with not very strong work experience on standalone applications (and I remember how much I detested to deal with issues related to loading resources :) )... and I fully understand your point on the number of users that might need this feature.

I would spend at most one hour to try to make it working with resources outside the HO.jar.
If I can't manage, I will focus on fixing the launch when predictions are into the JAR (and then we'll search a way to put them outside only in case there are consistent user requests for it).

**Comment by pierpaolocira** [comment]
@akasolace as additional information, outcomes of the tests I did:
* adding `prediction` in the HO.jar of the HO.app and launching the Mac .app --> the aforementioned NullPointerException 
* manually adding `prediction` in any possible location of HO.app (+ also at the same level of HO.app) and launching the Mac .app --> still original issue (ExceptionInInitialierError)
* manually adding `prediction` in the same level of HO.jar of the HO.app and launching the Mac .app --> this is included in the previous point, that means: still original issue (ExceptionInInitialierError)
* **BUT** manually adding `prediction` in the same level of HO.jar of the HO.app and launching via `java -jar HO.jar` --> this works and no exceptions are generated.

Please note on the last point: the ".app" Mac applications are just folders, so I went inside `HO.app/Contents/Java` and simply alunched the Jar.

This means that the application (regardless `prediction` is placed either beside or inside the HO.jar) works properly on MacOS. The issue happens if you launch HO with the .app  generated by the `macAppBundle`, but it doesn't if you launch the JAR taken from the content of the .app  generated by the `macAppBundle`.

In my view this means that:
* no specific way to load files should be put in place for Mac
* there is an issue with the packaging... but I cannot understand exactly what...

...of course I cannot exclude that if we put `prediction` into HO.jar and work to load files differently on Mac it will work... but I'm wondering what is the good way to be followed in order to keep the adequate level of maintainability...

By chance, do you remember exactly the latest working commit? I would like to analyse differences between them...

Additioanlly, I also tried to build with the most recent version (2.3.0 instead of 2.2.3) of `edu.sc.seis.macAppBundle` but nothing changed.

**Comment by akasolace** [comment]
@pierpaolocira I struggled to make it works the first time. 
At the moment I would stick with current approach

if you debug the app under macOs, can you see where it fails. 
Is it in `formulaFactorsFile = new File(getClass().getClassLoader().getResource("prediction/defaults.xml").getFile())`  ?

and why does it work in  `br = new BufferedReader(new InputStreamReader((loader.getResourceAsStream(newFilename))));`   I guess there is a subtle difference,

If I had access to Mac, I would step by step the 2 methods and see if I can spot a diference.

**Comment by pierpaolocira** [comment]
@akasolace yes, I noticed that... but the issue is that if I launch my IDE it works fine... and if I launch it via command line (java -jar) it works fine... and if I launch via command line the content of the ".app" file it continues to work fine!
The issue can be reproduced only when the application is launched as a .app... and I can't debug it.


By chance, do you remember exactly the latest working commit? I would like to analyse differences between them...
Or do you have any other idea?

**Comment by akasolace** [comment]
I guess the bug appeared approximtivly 2 months ago when I worked on HO/src/main/java/core/model/FormulaFactors.java 

Anyway, I believe the issue is that under Mac the ressource is within the JAR so it should be accessed via getResourceAsStream (as mentioned above). I see 2 approach:
    - 1) we can have a single method using getResourceAsStream  for all platform
    - 2) we use a specific piece of code under Mac OS

Again I think I would start looking at the piece of code reading datfile because this one works on all platform and has similar problematic, the files are within the JAR under macOS and outside the JAR in other platform.

**Comment by pierpaolocira** [comment]
@akasolace Thanks for the information.

I am not sure what you mentioned is the exact cause of the issue, because of one of my tests:

> manually adding prediction in the same level of HO.jar [beside and NOT inside] of the HO.app and launching via java -jar HO.jar --> this works and no exceptions are generated.

So in a situation where predictions is NOT in the jar, but in a directory at the same level --> the application runs properly on MacOX if launched directly via the jar (I mean, the JAR placed into the .app).
So **exactly the same files** work if you launch from the JAR and not if you launch the app containing the JAR.

Anyway I now have enough information to work on it. I hope to be able to give soon some positive updates to you.. ;)

**Comment by akasolace** [comment]
@pierpaolocira 

> > I am not sure what you mentioned is the exact cause of the issue, because of one of my tests:
> 
> manually adding prediction in the same level of HO.jar [beside and NOT inside] of the HO.app and launching via java -jar HO.jar --> this works and no exceptions are generated.

This confirms what I said, if ressouce is outsie of the jar it works, otherwise it needs to be called using `getResourceAsStream`.

Also as I mentioned, you can try to use only `getResourceAsStream` and I imagine it should work under all platform (as it is the case for dat file) otherise keep the code as is and simply use `getResourceAsStream` if platform is macOS.

**Comment by pierpaolocira** [comment]
@akasolace 

> This confirms what I said, if ressouce is outsie of the jar it works, otherwise it needs to be called using `getResourceAsStream`.

I was not telling that your approach shouldn't solve...

I would just keep track that the root cause of the issue is on the JavaLauncher (inside the .app)... because it seems not able to support loading of resources outside the jar.
Indeed the same folder structure lauched via "java -jar" works fine.
I think this is a limitation of the gradle plugin (indeed there is at least one fork -- but quite old -- able to define additional classpaths for loading resources).

Anyway, as agreed, I'll work on a simple solution and keep you posted.

**Comment by pierpaolocira** [comment]
Quick and very dirty...
Maybe we can have a unique utility class to load resources (that will take care to load a File o from the ClassLoader), in order to centralize all the calls.
If you agree, please create a new feature request and assign to me... it should be quite easy.

**Comment by akasolace** [comment]
closed by #203
